File: 6.6-react-agent/ghidra_scripts/export_analysis.py
# Ghidra Jython script. Run through analyzeHeadless with:
# -postScript export_analysis.py <output_dir> <max_functions>

from __future__ import print_function

from ghidra.app.decompiler import DecompInterface
from ghidra.program.util import DefinedDataIterator
from ghidra.util.task import ConsoleTaskMonitor

import json
import os
import codecs


def safe_text(value):
    try:
        return unicode(value)
    except NameError:
        return str(value)
    except Exception:
        return str(value)


def get_program_info(program):
    lang = program.getLanguage()
    compiler = program.getCompilerSpec()
    return {
        "name": safe_text(program.getName()),
        "executable_path": safe_text(program.getExecutablePath()),
        "language": safe_text(lang.getLanguageID()),
        "compiler": safe_text(compiler.getCompilerSpecID()),
        "image_base": safe_text(program.getImageBase()),
    }


def export_strings(program, max_items):
    values = []
    iterator = DefinedDataIterator.definedStrings(program)
    count = 0
    while iterator.hasNext() and count < max_items:
        data = iterator.next()
        values.append(
            {
                "address": safe_text(data.getAddress()),
                "value": safe_text(data.getValue()),
            }
        )
        count += 1
    return values


def export_imports(program):
    symbols = []
    symbol_table = program.getSymbolTable()
    iterator = symbol_table.getExternalSymbols()
    while iterator.hasNext():
        symbol = iterator.next()
        symbols.append(
            {
                "name": safe_text(symbol.getName(True)),
                "address": safe_text(symbol.getAddress()),
            }
        )
    return symbols


def collect_calls(program, function):
    listing = program.getListing()
    function_manager = program.getFunctionManager()
    calls = []
    instructions = listing.getInstructions(function.getBody(), True)
    while instructions.hasNext():
        instruction = instructions.next()
        refs = instruction.getReferencesFrom()
        for ref in refs:
            if ref.getReferenceType().isCall():
                target = ref.getToAddress()
                target_function = function_manager.getFunctionAt(target)
                calls.append(
                    {
                        "from": safe_text(instruction.getAddress()),
                        "to": safe_text(target),
                        "name": safe_text(target_function.getName(True))
                        if target_function
                        else safe_text(target),
                    }
                )
    return calls


def decompile_function(decompiler, function, monitor):
    try:
        result = decompiler.decompileFunction(function, 30, monitor)
        if result and result.decompileCompleted() and result.getDecompiledFunction():
            return safe_text(result.getDecompiledFunction().getC())
        if result:
            return "DECOMPILE_FAILED: " + safe_text(result.getErrorMessage())
    except Exception as exc:
        return "DECOMPILE_EXCEPTION: " + safe_text(exc)
    return "DECOMPILE_FAILED"


def export_functions(program, max_functions):
    function_manager = program.getFunctionManager()
    monitor = ConsoleTaskMonitor()
    decompiler = DecompInterface()
    decompiler.openProgram(program)
    functions = []
    count = 0
    iterator = function_manager.getFunctions(True)
    while iterator.hasNext() and count < max_functions:
        function = iterator.next()
        functions.append(
            {
                "entry": safe_text(function.getEntryPoint()),
                "name": safe_text(function.getName(True)),
                "signature": safe_text(function.getSignature()),
                "is_thunk": bool(function.isThunk()),
                "calls": collect_calls(program, function),
                "decompiled": decompile_function(decompiler, function, monitor),
            }
        )
        count += 1
    decompiler.dispose()
    return functions


def main():
    args = getScriptArgs()
    if len(args) < 1:
        raise Exception("export_analysis.py requires output_dir argument")
    output_dir = args[0]
    max_functions = int(args[1]) if len(args) > 1 else 200
    if not os.path.isdir(output_dir):
        os.makedirs(output_dir)

    program = currentProgram
    payload = {
        "program": get_program_info(program),
        "imports": export_imports(program),
        "strings": export_strings(program, 500),
        "functions": export_functions(program, max_functions),
    }

    output_path = os.path.join(output_dir, "analysis.json")
    with codecs.open(output_path, "w", "utf-8") as fh:
        json.dump(payload, fh, indent=2, ensure_ascii=False)
    print("Wrote " + output_path)


main()

File: 6.6-react-agent/ghidra_scripts/ExportAnalysis.java
import ghidra.app.script.GhidraScript;
import ghidra.app.decompiler.*;
import ghidra.program.model.address.Address;
import ghidra.program.model.listing.*;
import ghidra.program.model.symbol.*;
import ghidra.program.util.*;
import ghidra.util.task.*;

import java.io.*;
import java.nio.charset.StandardCharsets;
import java.nio.file.*;

public class ExportAnalysis extends GhidraScript {

    @Override
    public void run() throws Exception {
        String[] args = getScriptArgs();
        if (args.length < 1) {
            throw new IllegalArgumentException("export_analysis requires output_dir argument");
        }
        String outputDir = args[0];
        int maxFunctions = args.length > 1 ? Integer.parseInt(args[1]) : 200;

        Files.createDirectories(Paths.get(outputDir));

        Program program = getCurrentProgram();
        StringBuilder sb = new StringBuilder();
        sb.append("{\n");

        sb.append("  \"program\": ");
        appendProgramInfo(sb, program);
        sb.append(",\n");

        sb.append("  \"imports\": ");
        appendImports(sb, program);
        sb.append(",\n");

        sb.append("  \"strings\": ");
        appendStrings(sb, program, 500);
        sb.append(",\n");

        sb.append("  \"functions\": ");
        appendFunctions(sb, program, maxFunctions);
        sb.append("\n}\n");

        File outputFile = Paths.get(outputDir, "analysis.json").toFile();
        try (OutputStreamWriter writer = new OutputStreamWriter(
                new FileOutputStream(outputFile), StandardCharsets.UTF_8)) {
            writer.write(sb.toString());
        }
        println("Wrote " + outputFile.getAbsolutePath());
    }

    private void appendProgramInfo(StringBuilder sb, Program program) {
        sb.append("{\n");
        sb.append("    \"name\": ").append(jsonEscape(program.getName())).append(",\n");
        sb.append("    \"executable_path\": ").append(jsonEscape(program.getExecutablePath())).append(",\n");
        sb.append("    \"language\": ").append(jsonEscape(program.getLanguage().getLanguageID().getIdAsString())).append(",\n");
        sb.append("    \"compiler\": ").append(jsonEscape(program.getCompilerSpec().getCompilerSpecID().getIdAsString())).append(",\n");
        sb.append("    \"image_base\": ").append(jsonEscape(program.getImageBase().toString())).append("\n");
        sb.append("  }");
    }

    private void appendStrings(StringBuilder sb, Program program, int maxItems) {
        sb.append("[\n");
        DataIterator stringIter = program.getListing().getDefinedData(true);
        int count = 0;
        boolean first = true;
        while (stringIter.hasNext() && count < maxItems) {
            Data data = stringIter.next();
            if (!data.hasStringValue()) continue;
            if (!first) sb.append(",\n");
            first = false;
            sb.append("    {");
            sb.append("\"address\": ").append(jsonEscape(data.getAddress().toString()));
            sb.append(", \"value\": ").append(jsonEscape(data.getDefaultValueRepresentation()));
            sb.append("}");
            count++;
        }
        sb.append("\n  ]");
    }

    private void appendImports(StringBuilder sb, Program program) {
        sb.append("[\n");
        SymbolTable symbolTable = program.getSymbolTable();
        SymbolIterator iterator = symbolTable.getExternalSymbols();
        boolean first = true;
        while (iterator.hasNext()) {
            Symbol symbol = iterator.next();
            if (!first) sb.append(",\n");
            first = false;
            sb.append("    {");
            sb.append("\"name\": ").append(jsonEscape(symbol.getName(true)));
            sb.append(", \"address\": ").append(jsonEscape(symbol.getAddress().toString()));
            sb.append("}");
        }
        sb.append("\n  ]");
    }

    private void appendCallList(StringBuilder sb, Program program, Function function) {
        sb.append("\"calls\": [\n");
        Listing listing = program.getListing();
        FunctionManager functionManager = program.getFunctionManager();
        InstructionIterator instructions = listing.getInstructions(function.getBody(), true);
        boolean first = true;
        while (instructions.hasNext()) {
            Instruction instruction = instructions.next();
            Reference[] refs = instruction.getReferencesFrom();
            for (Reference ref : refs) {
                if (ref.getReferenceType().isCall()) {
                    if (!first) sb.append(",\n");
                    first = false;
                    Address target = ref.getToAddress();
                    Function targetFunction = functionManager.getFunctionAt(target);
                    sb.append("        {");
                    sb.append("\"from\": ").append(jsonEscape(instruction.getAddress().toString()));
                    sb.append(", \"to\": ").append(jsonEscape(target.toString()));
                    sb.append(", \"name\": ").append(jsonEscape(
                            targetFunction != null ? targetFunction.getName(true) : target.toString()));
                    sb.append("}");
                }
            }
        }
        sb.append("\n      ]");
    }

    private String decompileFunction(DecompInterface decompiler, Function function, ConsoleTaskMonitor monitor) {
        try {
            DecompileResults result = decompiler.decompileFunction(function, 30, monitor);
            if (result != null && result.decompileCompleted() && result.getDecompiledFunction() != null) {
                return jsonEscape(result.getDecompiledFunction().getC());
            }
            if (result != null) {
                return jsonEscape("DECOMPILE_FAILED: " + result.getErrorMessage());
            }
        } catch (Exception e) {
            return jsonEscape("DECOMPILE_EXCEPTION: " + e.getMessage());
        }
        return jsonEscape("DECOMPILE_FAILED");
    }

    private void appendFunctions(StringBuilder sb, Program program, int maxFunctions) {
        sb.append("[\n");
        FunctionManager functionManager = program.getFunctionManager();
        ConsoleTaskMonitor monitor = new ConsoleTaskMonitor();
        DecompInterface decompiler = new DecompInterface();
        decompiler.openProgram(program);
        int count = 0;
        boolean first = true;
        FunctionIterator iterator = functionManager.getFunctions(true);
        while (iterator.hasNext() && count < maxFunctions) {
            Function function = iterator.next();
            if (!first) sb.append(",\n");
            first = false;
            sb.append("    {\n");
            sb.append("      \"entry\": ").append(jsonEscape(function.getEntryPoint().toString())).append(",\n");
            sb.append("      \"name\": ").append(jsonEscape(function.getName(true))).append(",\n");
            sb.append("      \"signature\": ").append(jsonEscape(function.getSignature().toString())).append(",\n");
            sb.append("      \"is_thunk\": ").append(function.isThunk()).append(",\n");
            appendCallList(sb, program, function);
            sb.append(",\n");
            sb.append("      \"decompiled\": ").append(decompileFunction(decompiler, function, monitor)).append("\n");
            sb.append("    }");
            count++;
        }
        decompiler.dispose();
        sb.append("\n  ]");
    }

    private String jsonEscape(String s) {
        if (s == null) return "null";
        StringBuilder sb = new StringBuilder(s.length() + 4);
        sb.append('"');
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case '"':  sb.append("\\\""); break;
                case '\\': sb.append("\\\\"); break;
                case '\n': sb.append("\\n"); break;
                case '\r': sb.append("\\r"); break;
                case '\t': sb.append("\\t"); break;
                default:
                    if (c < 0x20) {
                        sb.append(String.format("\\u%04x", (int) c));
                    } else {
                        sb.append(c);
                    }
            }
        }
        sb.append('"');
        return sb.toString();
    }
}
